File: Week4/23.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Heap:
    def __init__(self):
        self.heap=[]
        self.size=0
    def push(self,val):
        self.heap.append(val)
        self.up(self.size)
        self.size+=1
    def pop(self):
        if self.size:
            res=self.heap[0]
            self.size-=1
            if self.size:
                self.heap[0]=self.heap.pop()
                self.down(0)
            else:
                self.heap.pop()
            return res
    def up(self,index):
        if index:
            parent=(index-1)//2
            if self.heap[parent]>self.heap[index]:
                self.swap(parent,index)
                self.up(parent)
    def down(self,index):
        left=index*2+1
        right=index*2+2
        Min=index
        if left<self.size and self.heap[left]<self.heap[Min]:
            Min=left
        if right<self.size and self.heap[right]<self.heap[Min]:
            Min=right
        if Min!=index:
            self.swap(Min,index)
            self.down(Min)
    def swap(self,p,q):
        self.heap[p],self.heap[q]=self.heap[q],self.heap[p]
    def isempty(self):
        return self.size==0
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        heap=Heap()
        for lis in lists:
            while lis:
                heap.push(lis.val)
                lis=lis.next
        s=ListNode(49)
        cur=s
        while not heap.isempty():
            cur.next=ListNode(heap.pop())
            cur=cur.next
        return s.next

File: Week4/23.java
import java.util.Collections;
 class MinHeap{
     List<ListNode> data;
     int size;
        public MinHeap(){
            size=0;
            data=new ArrayList<ListNode>();
        }
        public boolean isempty(){
            return size==0;
        }
        private void swap(int i,int j){
            Collections.swap(data, i, j);
        }
        private void up(int index){
            if(index>0){
                int parent=(index-1)/2;
                if(data.get(index).val<data.get(parent).val){
                    swap(index,parent);
                    up(parent);
                }
            }
        }
        private void down(int index){
            if(size!=0){
                int left=index*2+1;
                int right=index*2+2;
                int successor=index;
                if(right<size && data.get(right).val<data.get(successor).val){
                    successor=right;
                }
                if(left<size && data.get(left).val<data.get(successor).val){
                    successor=left;
                }
                if(successor!=index){
                    swap(successor, index);
                    down(successor);
                }
            }
        }
        public void add(ListNode node){
            data.add(node);
            up(size);
            size+=1;
        }
        public ListNode pop(){
            size-=1;
            ListNode res=data.get(0);
            swap(0, size);
            data.remove(size);
            down(0);
            return res;
        }
    }
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        MinHeap heap=new MinHeap();
        ListNode res=new ListNode();
        ListNode cur=res;
        for(ListNode l:lists){
            if(l!=null){
                heap.add(l);
            }
        }
        while(!heap.isempty()){
            ListNode m=heap.pop();
            cur.next=m;
            if(m.next!=null){
                heap.add(m.next);
            }
            cur=cur.next;
        }
        return res.next;
    }
}